### PR TITLE
Add support for progress notifiers

### PR DESCRIPTION
**Pull request by austinzheng**
Depends upon https://github.com/realm/realm-cocoa/pull/4478, https://github.com/realm/realm-object-store/pull/310

### CODE REVIEW

File: Realm/RLMSyncSession.mm [hunk 1]
```diff
@@ -24,10 +24,38 @@
...
+@implementation RLMProgressNotificationToken
+
+- (void)stop {
+    if (auto session = _session.lock()) {
+        session->unregister_progress_notifier(_token);
+    }
+}
+
+- (void)dealloc {
+    [self stop];
```
Comment (jpsim) [inline Realm/RLMSyncSession.mm]: This behavior is different than `RLMNotificationToken`:

```objc
- (void)dealloc {
    if (_realm || _block) {
        NSLog(@"RLMNotificationToken released without unregistering a notification. You must hold "
              @"on to the RLMNotificationToken returned from addNotificationBlock and call "
              @"-[RLMNotificationToken stop] when you no longer wish to receive RLMRealm notifications.");
    }
}
```

File: Realm/RLMSyncSession.mm [hunk 2]
```diff
@@ -24,10 +24,42 @@
...
+
+@implementation RLMProgressNotificationToken
+
+- (void)stop {
+    if (auto session = _session.lock()) {
+        session->unregister_progress_notifier(_token);
+    }
+}
 
+- (void)dealloc {
```
Comment (jpsim) [inline Realm/RLMSyncSession.mm]: This behavior is different than `RLMNotificationToken`:

```objc
- (void)dealloc {
    if (_realm || _block) {
        NSLog(@"RLMNotificationToken released without unregistering a notification. You must hold "
              @"on to the RLMNotificationToken returned from addNotificationBlock and call "
              @"-[RLMNotificationToken stop] when you no longer wish to receive RLMRealm notifications.");
    }
}
```

Comment (austinzheng) [inline Realm/RLMSyncSession.mm]: So the expected behavior is, upon dealloc, if `-stop` hasn't been called previously a `NSLog` warning message is fired?

Comment (jpsim) [inline Realm/RLMSyncSession.mm]: I'm not saying that should be the expected behavior, only that it currently diverges from `RLMNotificationToken`.

Comment (austinzheng) [inline Realm/RLMSyncSession.mm]: I changed it so it matches. Consistency is a good thing here.

**Comment by austinzheng** [comment]
I removed the use of `Optional<T>`, since @morten-krogh said he would change sync to call the notifier immediately after registration, meaning that we should always know the downloadable and uploadable bytes counts. (https://github.com/realm/product/issues/40#issuecomment-271144959)

**Pull request closed without merging**